hail/src/main/scala/is/hail/annotations/CodeOrdering.scala

```diff
@@ -68,294 +57,518 @@ object CodeOrdering {
-          Code(setup(i)(x, y),
-            mbequiv((m1, v1s(i)), (m2, v2s(i))))
-        }.foldRight[Code[Boolean]](const(true))(_ && _)
+      cb.define(Lout)
+      gteq
+    }
+
+    override def equivNonnull(cb:EmitCodeBuilder, x: PCode, y: PCode): Code[Boolean] = {
+      val (lhs, rhs) = setup(cb, x, y)
+      val Lout = CodeLabel()
+      val eq = cb.newLocal("cmp", true)
+
+      var i = 0
+      while (i < t1.size) {
+        val fldEq = fieldOrdering(i, CodeOrdering.Equiv(missingFieldsEqual))
+        val l = EmitCode.fromI(cb.emb) { cb => lhs.loadField(cb, i).typecast }
+        val r = EmitCode.fromI(cb.emb) { cb => rhs.loadField(cb, i).typecast }
+        cb.assign(eq, fldEq(cb, l, r))
+        cb.ifx(!eq, cb.goto(Lout))
+        i += 1
       }
+
+      cb.define(Lout)
+      eq
+    }
   }
 
   def iterableOrdering(t1: PArray, t2: PArray, mb: EmitMethodBuilder[_]): CodeOrdering = new CodeOrdering {
-    type T = Long
-    val lord: CodeOrdering = PInt32().codeOrdering(mb)
-    val ord: CodeOrdering = t1.elementType.codeOrdering(mb, t2.elementType)
-    val len1: LocalRef[Int] = mb.newLocal[Int]()
-    val len2: LocalRef[Int] = mb.newLocal[Int]()
-    val lim: LocalRef[Int] = mb.newLocal[Int]()
-    val i: LocalRef[Int] = mb.newLocal[Int]()
-    val m1: LocalRef[Boolean] = mb.newLocal[Boolean]()
-    val v1: LocalRef[ord.T] = mb.newLocal()(typeToTypeInfo(t1.elementType)).asInstanceOf[LocalRef[ord.T]]
-    val m2: LocalRef[Boolean] = mb.newLocal[Boolean]()
-    val v2: LocalRef[ord.T] = mb.newLocal()(typeToTypeInfo(t2.elementType)).asInstanceOf[LocalRef[ord.T]]
-    val eq: LocalRef[Boolean] = mb.newLocal[Boolean]()
-
-    def loop(cmp: Code[Unit], loopCond: Code[Boolean])
-      (x: Code[Long], y: Code[Long]): Code[Unit] = {
-      Code.memoize(x, "cord_iter_ord_x", y, "cord_iter_ord_y") { (x, y) =>
-        Code(
-          i := 0,
-          len1 := t1.loadLength(x),
-          len2 := t2.loadLength(y),
-          lim := (len1 < len2).mux(len1, len2),
-          Code.whileLoop(loopCond && i < lim,
-            m1 := t1.isElementMissing(x, i),
-            v1.storeAny(Region.loadIRIntermediate(t1.elementType)(t1.elementOffset(x, len1, i))),
-            m2 := t2.isElementMissing(y, i),
-            v2.storeAny(Region.loadIRIntermediate(t2.elementType)(t2.elementOffset(y, len2, i))),
-            cmp, i += 1))
-      }
+    private[this] def setup(cb: EmitCodeBuilder, lhs: PCode, rhs: PCode): (PIndexableValue, PIndexableValue) = {
+      val lhsv = lhs.asIndexable.memoize(cb, "container_ord_lhs")
+      val rhsv = rhs.asIndexable.memoize(cb, "container_ord_rhs")
+      lhsv -> rhsv
     }
 
-    override def compareNonnull(x: Code[Long], y: Code[Long]): Code[Int] = {
-      val mbcmp = mb.getCodeOrdering(t1.elementType, t2.elementType, CodeOrdering.Compare())
-      val cmp = mb.newLocal[Int]()
+    private[this] def loop(cb: EmitCodeBuilder, lhs: PIndexableValue, rhs: PIndexableValue)(
+      f: (EmitCode, EmitCode) => Unit
```

Comment:
Here we take `cb` as a parameter to `loop`, and `f` doesn't take a `cb` -- it appends to another cb, implicitly assuming they're the same. I think it's much safer to always have functions like `f` here take a cb, so it's easier to locally interpret the code generated.